ClientSyncComponent/ClientSyncComponent.Client/Pages/ClientSyncComponent.razor.cs

```diff
@@ -0,0 +1,170 @@
+
+        List<string> ReceivedValues { get; set; } = new List<string>();
+
+        Timer Timer { get; set; }
+
+        protected override Task OnParametersSetAsync()
+        {
+            TableClient = new TableClient(TableSASUrl);
+            return base.OnParametersSetAsync();
+        }
+
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            if (firstRender)
+            {
+                await JSRuntime.InvokeVoidAsync("registerDotNet", DotNetObjectReference.Create(this));
+            }
+
+            await base.OnAfterRenderAsync(firstRender);
+        }
+
+        private async void OnTimer(object? state)
+        {
+            await SyncAsync();
+            string visibility = await JSRuntime.InvokeAsync<string>("getVisibility");
+            Console.WriteLine("Timer ticked. Visibility is " + visibility);
+
+            int newInterval = CalculateInterval();
+            Timer.Change(newInterval, newInterval);
+        }
+
+        /// <summary>
+        /// Heuristic to minimize load depending on how active a puzzle is
+        /// </summary>
+        /// <returns>Interval in milliseconds to poll storage at</returns>
+        private int CalculateInterval()
+        {
+            // todo: get a multiplier or override from the server for central updates?
+            TimeSpan timeSinceUpdate = DateTimeOffset.UtcNow - LastSyncUtc;
+            if (timeSinceUpdate.TotalSeconds < 120)
+            {
+                return 1000;
+            }
+            else if (timeSinceUpdate.TotalMinutes < 10)
+            {
+                return 2000;
+            }
+            else if (timeSinceUpdate.TotalHours < 1)
+            {
+                return 5000;
+            }
+            else if (timeSinceUpdate.TotalHours < 6)
+            {
+                return 10000;
+            }
+            else
+            {
+                return 60000;
+            }
+        }
+
+        [JSInvokable]
+        public void VisibilityChanged(string visibility)
+        {
+            Console.WriteLine($".NET Visibility changed to {visibility}");
+            if (visibility == "visible")
+            {
+                Timer.Change(0, CalculateInterval());
+            }
+            else
+            {
+                Timer.Change(Timeout.Infinite, Timeout.Infinite);
+            }
+        }
+
+        [JSInvokable]
+        public void OnSyncablePuzzleLoaded()
+        {
+            Console.WriteLine("Syncable puzzle loaded, starting sync timer");
+            if (Timer == null)
+            {
+                Timer = new Timer(OnTimer, null, 1000, 1000);
+            }
+        }
+
+        [JSInvokable]
+        public async void OnPuzzleChangedAsync(JsPuzzleChange[] puzzleChanges)
+        {
+            List<TableTransactionAction> actions = new List<TableTransactionAction>(puzzleChanges.Length);
+
+            foreach (JsPuzzleChange change in puzzleChanges)
```

Comment:
Strongly suggest using `TableClient.SubmitTransactionAsync` instead of multiple `TableClient.UpsertEntityAsync` calls, as it is atomic when there are multiple changes. I checked, and it supports upserts.

Reply:
This doesn't seem to be working, I get a 403 for batch requests